### PR TITLE
Adds new samplers and benchmarks

### PR DESCRIPTION
**Pull request by adriancole**
This introduces two options for trace sampling:

* BoundaryTraceIdSampler - For high-volume sites, based on Finagle
* CountingTraceIdSampler - Trades lower performance for high accuracy

These include benchmarks and both work well, although the boundary
one is an order of magnitude faster under contention.

Here's the JMH comparison from my macbook pro and 4 threads:

```
SamplerBenchmarks.sampler_boundary            -9223372036854775808  thrpt   15  875.378 ± 28.245  ops/us
SamplerBenchmarks.sampler_boundary             1234567890987654321  thrpt   15  819.325 ± 81.720  ops/us
SamplerBenchmarks.sampler_counting            -9223372036854775808  thrpt   15   22.777 ±  0.540  ops/us
SamplerBenchmarks.sampler_counting             1234567890987654321  thrpt   15   22.223 ±  0.695  ops/us
```

CountingSampler is a better default, as support questions have come up with boundary tracing. A lot of people don't have much traffic and they can be surprised to find nothing being sampled at all!

This counting sampler is a thread-safe version of the old one, which gives an exact sample rate, per 100 decisions. The decisions are shuffled, so people should get fair decisions even before the 100 interval is over. It can do several samples per microsecond under contention, which is good enough for people getting started.

Basically, by switching to this, people can have a better experience, and save the tricky stuff for later.

### CODE REVIEW

**Comment by adriancole** [comment]
CountingSampler is a better default, as support questions have come up with boundary tracing. A lot of people don't have much traffic and they can be surprised to find nothing being sampled at all!

This counting sampler is a thread-safe version of the old one, which gives an *exact* sample rate. It can do several samples per microsecond under contention, which is good enough for people getting started. 

Basically, by switching to this, people can have a better experience, and save the tricky stuff for later.

**Comment by adriancole** [comment]
@kristofa @eirslett if you don't mind.. ptal!

File: brave-core/src/test/java/com/github/kristofa/brave/BoundarySamplerTest.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+import static org.assertj.core.data.Percentage.withPercentage;
+
+public class BoundarySamplerTest extends SamplerTest {
+  @Override Sampler newSampler(float rate) {
+    return BoundarySampler.create(rate);
+  }
+
+  @Override Percentage expectedErrorRate() {
+    return withPercentage(10);
```
Comment (adriancole) [inline brave-core/src/test/java/com/github/kristofa/brave/BoundarySamplerTest.java]: note that usually this is good in 3-5% range over 100K ids. However, sometimes it isn't, so the threshold is high to not flake travis.

File: brave-core/src/test/java/com/github/kristofa/brave/CountingSamplerTest.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+import static org.assertj.core.data.Percentage.withPercentage;
+
+public class CountingSamplerTest extends SamplerTest {
+  @Override Sampler newSampler(float rate) {
+    return CountingSampler.create(rate);
+  }
+
+  @Override Percentage expectedErrorRate() {
+    return withPercentage(0);
```
Comment (adriancole) [inline brave-core/src/test/java/com/github/kristofa/brave/CountingSamplerTest.java]: note: this is exact, even during the parallel test.

**Comment by eirslett** [comment]
LGTM!
